# 6.1 Sourcing Open Data

## Contents list:

### Data Wrangling Procedures & Data consistency checks
#### 01 Importing libraries
#### 02 Importing data
#### 03 Data Wrangling Procedures
#### 04 Data consistency checks


### 01 Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

### 02 Importing data

In [2]:
# Project folder path as string
path = r'C:\Users\User\Desktop\Chocolate Bar Ratings'

In [3]:
# Import flavors_of_cacao.csv data set using the os library
df_c= pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'flavors_of_cacao.csv'), index_col = False)

In [4]:
# Checking data info
df_c.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1795 entries, 0 to 1794
Data columns (total 9 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Company 
(Maker-if known)         1795 non-null   object 
 1   Specific Bean Origin
or Bar Name  1795 non-null   object 
 2   REF                               1795 non-null   int64  
 3   Review
Date                       1795 non-null   int64  
 4   Cocoa
Percent                     1795 non-null   object 
 5   Company
Location                  1795 non-null   object 
 6   Rating                            1795 non-null   float64
 7   Bean
Type                         1794 non-null   object 
 8   Broad Bean
Origin                 1794 non-null   object 
dtypes: float64(1), int64(2), object(6)
memory usage: 126.3+ KB


In [5]:
# Checking the number of rows and columns
df_c.shape

(1795, 9)

In [6]:
df_c.head()

Company \n(Maker-if known) Specific Bean Origin\nor Bar Name   REF  \
0                   A. Morin                       Agua Grande  1876   
1                   A. Morin                             Kpime  1676   
2                   A. Morin                            Atsane  1676   
3                   A. Morin                             Akata  1680   
4                   A. Morin                            Quilla  1704   

   Review\nDate Cocoa\nPercent Company\nLocation  Rating Bean\nType  \
0          2016            63%            France    3.75              
1          2015            70%            France    2.75              
2          2015            70%            France    3.00              
3          2015            70%            France    3.50              
4          2015            70%            France    3.50              

  Broad Bean\nOrigin  
0           Sao Tome  
1               Togo  
2               Togo  
3               Togo  
4               Peru

### 03 Data Wrangling Procedures

In [7]:
# Checking column names
df_c.columns

Index(['Company \n(Maker-if known)', 'Specific Bean Origin\nor Bar Name',
       'REF', 'Review\nDate', 'Cocoa\nPercent', 'Company\nLocation', 'Rating',
       'Bean\nType', 'Broad Bean\nOrigin'],
      dtype='object')

In [8]:
# Renaming columns
df_c.rename(columns = {'Company\xa0\n(Maker-if known)':'company',
                      'Specific Bean Origin\nor Bar Name':'specific_bean_origin_or_bar_name',
                      'REF':'reference_number',
                      'Review\nDate':'review_date',
                      'Cocoa\nPercent':'cocoa_percent',
                      'Company\nLocation':'company_location',
                      'Rating':'rating',
                      'Bean\nType':'bean_type',
                      'Broad Bean\nOrigin':'broad_bean_origin'}, inplace = True)

In [9]:
# Re-checking column names
df_c.columns

Index(['company', 'specific_bean_origin_or_bar_name', 'reference_number',
       'review_date', 'cocoa_percent', 'company_location', 'rating',
       'bean_type', 'broad_bean_origin'],
      dtype='object')

In [10]:
# Checking data types
df_c.dtypes

company                              object
specific_bean_origin_or_bar_name     object
reference_number                      int64
review_date                           int64
cocoa_percent                        object
company_location                     object
rating                              float64
bean_type                            object
broad_bean_origin                    object
dtype: object

In [11]:
# Basic statistic
df_c.describe()

reference_number  review_date       rating
count       1795.000000  1795.000000  1795.000000
mean        1035.904735  2012.325348     3.185933
std          552.886365     2.927210     0.478062
min            5.000000  2006.000000     1.000000
25%          576.000000  2010.000000     2.875000
50%         1069.000000  2013.000000     3.250000
75%         1502.000000  2015.000000     3.500000
max         1952.000000  2017.000000     5.000000

- The variable 'cocoa_percent' is a numeric variable expressed as a percentage, so it should be in the table above. In the output (10) of the previous function, it is visible that the variable 'cocoa_percent' has an object data type, that data type should be changed to float.

In [12]:
# Changing data type of 'cocoa_percent' variable from object into float because it is percentage
df_c['cocoa_percent'] = df_c['cocoa_percent'].str.rstrip("%").astype(float)/100

In [13]:
# Re-checking data types
df_c.dtypes

company                              object
specific_bean_origin_or_bar_name     object
reference_number                      int64
review_date                           int64
cocoa_percent                       float64
company_location                     object
rating                              float64
bean_type                            object
broad_bean_origin                    object
dtype: object

In [14]:
# Re-run basic statistic
df_c.describe()

reference_number  review_date  cocoa_percent       rating
count       1795.000000  1795.000000    1795.000000  1795.000000
mean        1035.904735  2012.325348       0.716983     3.185933
std          552.886365     2.927210       0.063231     0.478062
min            5.000000  2006.000000       0.420000     1.000000
25%          576.000000  2010.000000       0.700000     2.875000
50%         1069.000000  2013.000000       0.700000     3.250000
75%         1502.000000  2015.000000       0.750000     3.500000
max         1952.000000  2017.000000       1.000000     5.000000

- Now describe-table contain 'cocoa_percent' column.

### 04 Data consistency checks

In [15]:
# Missing values - finding missing values
df_c.isnull().sum()

company                             0
specific_bean_origin_or_bar_name    0
reference_number                    0
review_date                         0
cocoa_percent                       0
company_location                    0
rating                              0
bean_type                           1
broad_bean_origin                   1
dtype: int64

In [16]:
df_c['bean_type'].value_counts()

                            887
Trinitario                  419
Criollo                     153
Forastero                    87
Forastero (Nacional)         52
Blend                        41
Criollo, Trinitario          39
Forastero (Arriba)           37
Criollo (Porcelana)          10
Trinitario, Criollo           9
Forastero (Parazinho)         8
Forastero (Arriba) ASS        6
Nacional (Arriba)             3
Matina                        3
EET                           3
Beniano                       3
Criollo (Ocumare 61)          2
Trinitario, Forastero         2
Trinitario (85% Criollo)      2
Forastero (Catongo)           2
Criollo, Forastero            2
Amazon, ICS                   2
Criollo (Amarru)              2
Amazon mix                    2
Nacional                      2
Criollo (Ocumare)             1
Criollo (Ocumare 67)          1
Trinitario, TCGA              1
Trinitario (Amelonado)        1
Trinitario, Nacional          1
Forastero (Amelonado)         1
Foraster

- 'bean_type' variable contains 887 blank cells.

In [17]:
# Checking for empty cell or NaN
df_c.isnull().values.any()

True

In [18]:
# Checking for unique values 
df_c['bean_type'].unique()

array(['\xa0', 'Criollo', 'Trinitario', 'Forastero (Arriba)', 'Forastero',
       'Forastero (Nacional)', 'Criollo, Trinitario',
       'Criollo (Porcelana)', 'Blend', 'Trinitario (85% Criollo)',
       'Forastero (Catongo)', 'Forastero (Parazinho)',
       'Trinitario, Criollo', 'CCN51', 'Criollo (Ocumare)', 'Nacional',
       'Criollo (Ocumare 61)', 'Criollo (Ocumare 77)',
       'Criollo (Ocumare 67)', 'Criollo (Wild)', 'Beniano', 'Amazon mix',
       'Trinitario, Forastero', 'Forastero (Arriba) ASS', 'Criollo, +',
       'Amazon', 'Amazon, ICS', 'EET', 'Blend-Forastero,Criollo',
       'Trinitario (Scavina)', 'Criollo, Forastero', 'Matina',
       'Forastero(Arriba, CCN)', 'Nacional (Arriba)',
       'Forastero (Arriba) ASSS', 'Forastero, Trinitario',
       'Forastero (Amelonado)', nan, 'Trinitario, Nacional',
       'Trinitario (Amelonado)', 'Trinitario, TCGA', 'Criollo (Amarru)'],
      dtype=object)

- There are entries '\xa0' and 'nan'.

In [19]:
# Replacing '\xa0'and 'nan' entry with nan
df_c=df_c.replace('\xa0', np. nan)\
         .replace('nan', np. nan)

In [20]:
# Re-checking for missing values - finding missing values
df_c.isnull().sum()

company                               0
specific_bean_origin_or_bar_name      0
reference_number                      0
review_date                           0
cocoa_percent                         0
company_location                      0
rating                                0
bean_type                           888
broad_bean_origin                    74
dtype: int64

In [21]:
# Percentage of missing values
Missing_values=df_c.isnull().sum()/len(df_c)

In [22]:
Missing_values

company                             0.000000
specific_bean_origin_or_bar_name    0.000000
reference_number                    0.000000
review_date                         0.000000
cocoa_percent                       0.000000
company_location                    0.000000
rating                              0.000000
bean_type                           0.494708
broad_bean_origin                   0.041226
dtype: float64

- Percentage of missing values for 'bean type' = 49.4708 %. Since 49.4708 > 5 , these values should be inserted by one of the imputation methods or drop a column based on the percentage of missing values.
- Percentage of missing values for 'broad_bean_origin' = 4.1226 %. Since 4.1226 < 5, we can leave these values as they are or drop only these values.

In [23]:
# Dropping the bean_type column from df_c dataframe
df_c=df_c.drop(columns = ['bean_type'])

In [24]:
# Dropping only missing values
df_c=df_c.dropna(axis=0)

In [25]:
# Re-checking for missing values - finding missing values
df_c.isnull().sum()

company                             0
specific_bean_origin_or_bar_name    0
reference_number                    0
review_date                         0
cocoa_percent                       0
company_location                    0
rating                              0
broad_bean_origin                   0
dtype: int64

In [26]:
# Checking for number of rows and columns
df_c.shape

(1721, 8)

In [27]:
# The geographical feature is a significant criterion for Achievement 6
# Checking for unique values 
df_c['company_location'].unique()

array(['France', 'U.S.A.', 'Fiji', 'Ecuador', 'Mexico', 'Switzerland',
       'Netherlands', 'Spain', 'Peru', 'Canada', 'Italy', 'Brazil',
       'U.K.', 'Australia', 'Wales', 'Belgium', 'Germany', 'Russia',
       'Puerto Rico', 'Venezuela', 'Colombia', 'Japan', 'New Zealand',
       'Costa Rica', 'South Korea', 'Amsterdam', 'Scotland', 'Martinique',
       'Sao Tome', 'Argentina', 'Guatemala', 'South Africa', 'Bolivia',
       'St. Lucia', 'Portugal', 'Singapore', 'Denmark', 'Vietnam',
       'Grenada', 'Israel', 'India', 'Czech Republic',
       'Domincan Republic', 'Finland', 'Madagascar', 'Philippines',
       'Sweden', 'Poland', 'Austria', 'Honduras', 'Nicaragua',
       'Lithuania', 'Niacragua', 'Chile', 'Ghana', 'Iceland', 'Eucador',
       'Hungary', 'Suriname', 'Ireland'], dtype=object)

In [28]:
# Changing abbreviations to the full name and correcting typos
df_c['company_location']=df_c['company_location'].replace('Eucador', 'Ecuador')\
                                                 .replace('U.S.A.', 'United States of America')\
                                                 .replace('U.K.', 'United Kingdom')\
                                                 .replace('Amsterdam', 'Netherlands')\
                                                 .replace('Domincan Republic', 'Dominican Republic')\
                                                 .replace('Niacragua', 'Nicaragua')

In [29]:
# Checking values 
df_c['company_location'].value_counts()

United States of America    737
France                      144
Canada                      117
United Kingdom               95
Ecuador                      55
Italy                        54
Australia                    46
Belgium                      37
Germany                      35
Switzerland                  35
Austria                      23
Colombia                     23
Hungary                      22
Spain                        21
Venezuela                    20
Brazil                       17
Peru                         17
Madagascar                   17
Japan                        17
New Zealand                  17
Denmark                      15
Vietnam                      11
Scotland                     10
Guatemala                    10
Israel                        9
Argentina                     9
Costa Rica                    9
Netherlands                   8
Poland                        7
Lithuania                     6
Nicaragua                     6
Honduras

In [30]:
# Checking for unique values 
df_c['review_date'].unique()

array([2016, 2015, 2014, 2013, 2012, 2011, 2009, 2010, 2017, 2008, 2007,
       2006], dtype=int64)

In [31]:
# Checking for unique values 
df_c['cocoa_percent'].unique()

array([0.63 , 0.7  , 0.6  , 0.8  , 0.88 , 0.72 , 0.55 , 0.75 , 0.65 ,
       0.85 , 0.73 , 0.64 , 0.66 , 0.68 , 0.5  , 1.   , 0.77 , 0.9  ,
       0.71 , 0.83 , 0.78 , 0.74 , 0.76 , 0.86 , 0.82 , 0.69 , 0.91 ,
       0.42 , 0.61 , 0.735, 0.62 , 0.67 , 0.58 , 0.605, 0.79 , 0.81 ,
       0.57 , 0.725, 0.56 , 0.46 , 0.89 , 0.84 , 0.53 , 0.87 , 0.99 ])

In [32]:
# Checking for unique values 
df_c['rating'].unique()

array([3.75, 2.75, 3.  , 3.5 , 4.  , 3.25, 2.5 , 5.  , 1.75, 1.5 , 2.25,
       2.  , 1.  ])

In [33]:
# Checking for mixed-type data
for col in df_c.columns.tolist():
  weird = (df_c[[col]].applymap(type) != df_c[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_c[weird]) > 0:
    print (col)


- No mixed-type data found.

In [34]:
# Duplicates - finding duplicates
df_c1 = df_c[df_c.duplicated()]

In [35]:
# Checking for duplicates
df_c1

Empty DataFrame
Columns: [company, specific_bean_origin_or_bar_name, reference_number, review_date, cocoa_percent, company_location, rating, broad_bean_origin]
Index: []

In [36]:
df_c1.shape

(0, 8)

- No duplicates found.

In [37]:
# Exporting df_c dataframe as "chocolate_checked.csv" in "Prepared Data" folder.
df_c.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'chocolate_checked.csv')) 
